### PR TITLE
[AST, Sema] Replace HoleType with PlaceholderType

### PR DESCRIPTION
**Pull request by Jumhyn**
This is a first pass at an implementation for [placeholder types](https://forums.swift.org/t/placeholder-types/41329). Not yet ready for prime-time but much of the functionality is there!

### CODE REVIEW

File: lib/Sema/TypeChecker.cpp [hunk 1]
```diff
@@ -388,6 +388,10 @@ Type swift::performTypeResolution(TypeRepr *TyR, ASTContext &Ctx,
         // FIXME: Don't let unbound generic types escape type resolution.
         // For now, just return the unbound generic type.
         return unboundTy;
+      }, /*placeholderHandler*/ [&]() {
+        // FIXME: Don't let placeholder types escape type resolution.
+        // For now, just return the placeholder type.
+        return Ctx.ThePlaceholderType;
```
Comment (CodaFi) [inline lib/Sema/TypeChecker.cpp:394]: Maybe not something to do here specifically, but I was really hoping we could use placeholder types to make UnboundGenericType obsolete. Wherever type resolution forms one, it could instead build a bound generic type full of placeholders. With that in mind, having them escape unopened is not so bad.

**Review by CodaFi — COMMENTED** [review]
A fantastic first go. I can only speak to the type AST and type resolution bits though.

File: lib/Sema/TypeChecker.cpp (hunk 1 above)
Comment (Jumhyn) [reply lib/Sema/TypeChecker.cpp:394]: Yeah this was on my mind as well, especially since one of the goals of the proposal is to have `UnboundGeneric` be semantically equivalent to `UnboundGeneric<_, ..., _>`. Obsoleting `UnboundGenericType` would really help validate that approach.

It's been a bit since I worked on this but IIRC I started down that path, and then realized there was enough complexity (and bits that I didn't quite understand) in `ConstraintSystem::openUnboundGenericType` that it was going to turn into a bit of a rabbit hole, so I punted on that in favor of getting the basic functionality working.

Thank you for taking a look at this!

**Comment by CodaFi** [comment]
@Jumhyn @slavapestov and I are really excited to see this go in sooner rather than later. If you could separate the parsing code from everything else, we could get this merged immediately. Even if the proposal is not accepted, the re-model in here is extremely valuable.

**Comment by Jumhyn** [comment]
@CodaFi Thank you for the heads up. I'll work on separating those parts out. Could you provide just a bit more specificity on which aspects you'll want to merge right away? (E.g., do you want just the minimum to support the changes made in _ParseType.cpp_, or do you want basically everything here excluding the changes to `TypeResolution::forContextual` that actually generate the type variables for placeholder types?) 

Just want to make sure we're on the same page!

**Comment by CodaFi** [comment]
Let's take the latter approach - basically everything that doesn't touch the surface syntax can come along for the ride.

**Comment by CodaFi** [comment]
That includes all the type checker changes, type resolution changes, AST changes, etc. Just not the parser bits.

**Comment by xedin** [comment]
@Jumhyn Why not use `HoleType` as an explicit placeholder the same way we use it as implicit one for diagnostics? It's already has originating type variable so it could be used to suggest possible types.

**Comment by CodaFi** [comment]
> Why not use HoleType as an explicit placeholder the same way we use it as implicit one for diagnostics?

Isn't `HoleType` a contextual type and `PlaceholderType` an interface type? You can't create a `HoleType` without a type variable for example. We could certainly open a `PlaceholderType` as a `HoleType`.

**Comment by xedin** [comment]
@CodaFi I would be great if we could unify three things here - HoleType, PlaceholderType and UnresolvedType - all three effectively aim to represent the same thing... I think we can rename HoleType into a PlaceholderType if that fits better and use it for unknown types in constraint system and AST - currently it's possible to use HoleType for type variable, declaration type and dependent member type.

**Comment by Jumhyn** [comment]
> That includes all the type checker changes, type resolution changes, AST changes, etc. Just not the parser bits.

Ahh I had interpreted your suggestion to "separate the parsing code" the other way around. 🤦‍♂️ That makes way more sense, thank you. Will do!

> I think we can rename HoleType into a PlaceholderType if that fits better and use it for unknown types in constraint system and AST.

So under this model we'd want to add `PlaceholderTypeRepr` as another potential `Originator` for a `HoleType`, and then type resolution in the constraint system context would turn `HoleType`s into `TypeVariableType`s, which would get turned back into `HoleType`s if they could not be solved. Does that sound right?

**Comment by xedin** [comment]
@Jumhyn Currently `HoleType` is a replacement for a type variable since we need all type variables to be bound to some type before valid solution could be formed. My suggestion would be:
- If placeholder arrives from a declaration - turn it into a type variable and record in the locator that type variable has originated from a placeholder (so that finalization could replace type variables back into originating placeholders if they weren't bound);
- if type variable couldn't be determined bind it to a placeholder to form a valid solution (place holder becomes a replacements for `HoleType`) in diagnostic mode;
- If solution has a PlaceholderType in it - use it in the AST instead of using `UnresolvedType`, currently `HoleType` are turned into `UnresolvedType`s.

**Comment by Jumhyn** [comment]
@CodaFi @xedin Sorry for the delay—this turned out to be a bit of a busy week for me. Hoping to get to this over the weekend!

**Comment by xedin** [comment]
No worries!

**Comment by Jumhyn** [comment]
Hey @xedin @CodaFi @hborla, please take a look at the latest and let me know your thoughts. I've merged `HoleType` and `PlaceholderType` under the `PlaceholderType` name, basically implementing the first two items from your last comment, @xedin. If this seems like the right direction I can move forward with merging `UnresolvedType` and `PlaceholderType` as well, either in this PR or a follow-up.

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -115,44 +116,44 @@ class RecursiveTypeProperties {
...
     /// This type expression contains a DynamicSelf type.
-    HasDynamicSelf       = 0x80,
+    HasDynamicSelf = 0x80,
 
     /// This type contains an Error type.
-    HasError             = 0x100,
+    HasError = 0x100,
 
     /// This type contains a DependentMemberType.
-    HasDependentMember   = 0x200,
```
Comment (xedin) [inline include/swift/AST/Types.h:147]: Just a note - making unrelated whitespace changes like this sometimes complicates the review since it's easy to miss value changes...

File: include/swift/Sema/CSBindings.h [hunk 3]
```diff
@@ -279,26 +279,26 @@ struct PotentialBindings {
   bool isPotentiallyIncomplete() const;
 
   /// If this type variable doesn't have any viable bindings, or
-  /// if there is only one binding and it's a hole type, consider
-  /// this type variable to be a hole in a constraint system
-  /// regardless of where hole type originated.
-  bool isHole() const {
-    if (isDirectHole())
+  /// if there is only one binding and it's a placeholder type, consider
```
Comment (xedin) [inline include/swift/Sema/CSBindings.h:282]: @Jumhyn @hborla What do you think about keeping "hole" terminology in the constraint system? My personal preference would be to keep calling "untyped" places in constraint system holes and call them placeholders in AST.

**Review by xedin — COMMENTED** [review]
Looks good to me so far!

File: include/swift/Sema/CSBindings.h (hunk 3 above)
Comment (hborla) [reply include/swift/Sema/CSBindings.h:282]: I agree - I like the term "hole" for the constraint system because "placeholder" implies that it can be replaced, but if we bind to a placeholder in the constraint system, that means it solver could not replace it with a real type.

File: include/swift/AST/Types.h (hunk 2 above)
Comment (Jumhyn) [reply include/swift/AST/Types.h:147]: Ah, that must have been `clang-format`, I didn't make any intentional whitespace changes here. I'll revert!

File: include/swift/Sema/CSBindings.h (hunk 3 above)
Comment (Jumhyn) [reply include/swift/Sema/CSBindings.h:282]: Sounds good to me, I don't have super strong feelings either way! So we'd keep the type name as `PlaceholderType`, but all the constraint system methods (such as `isDirectHole()`) stay as is?

Comment (xedin) [reply include/swift/Sema/CSBindings.h:282]: Yes, that’s the idea.

Comment (xedin) [reply include/swift/Sema/CSBindings.h:282]: Effectively we are just utilizing a placeholder in a solution and AST to represent places which couldn’t be inferred.

Comment (Jumhyn) [reply include/swift/Sema/CSBindings.h:282]: @hborla @xedin Thoughts on `TVO_CanBindToPlaceholder` vs. `TVO_CanBindToHole`? This is the only one in the constraint system that seems to me like the use of "hole" is perhaps more confusing than "placeholder", since it most literally represents whether we will bind the type var to a `PlaceholderType` or not. But I'm good with whatever y'all think is best!

Comment (xedin) [reply include/swift/Sema/CSBindings.h:282]: I think we should leave it as hole since it would be consistent with the rest.

**Comment by Jumhyn** [comment]
@xedin @hborla Have updated to restore the "hole" terminology throughout the constraint system except where we are actually referring to the `PlaceholderType` type directly. Let me know if y'all have further thoughts!

**Comment by Jumhyn** [comment]
Hey @xedin @hborla @CodaFi, just wanted to check in on the status here—is there further work this needs to be merge-able?

**Review by CodaFi — APPROVED** [review]
Holly and Pavel will have more to say. But LGTM

File: include/swift/Sema/ConstraintSystem.h [hunk 4]
```diff
@@ -5024,6 +5024,24 @@ class OpenUnboundGenericType {
   }
 };
 
+class HandlePlaceholderType {
+  ConstraintSystem &cs;
+  const ConstraintLocatorBuilder &locator;
```
Comment (xedin) [inline include/swift/Sema/ConstraintSystem.h:5029]: I'd suggest we allocate a locator here instead of referencing a builder since we'd have to build a locator per placeholder anyway.

**Review by xedin — APPROVED** [review]
LGTM!

File: include/swift/Sema/ConstraintSystem.h (hunk 4 above)
Comment (Jumhyn) [reply include/swift/Sema/ConstraintSystem.h:5029]: Gotcha, will update!

**Comment by Jumhyn** [comment]
@xedin Updated to allocate a `ConstraintLocator` upon construction of `HandlePlaceholderType`! Mind giving this a test run for me?

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by Jumhyn** [comment]
Thank you @hborla!

**Comment by hborla** [comment]
@Jumhyn No problem! If you'd like, I think it's time for you to [request commit access](https://swift.org/contributing/#commit-access) 🙂

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by Jumhyn** [comment]
@xedin Do these failures in the release source compat run seem related to you? What's the best way for me to investigate those failures locally?

**Comment by xedin** [comment]
I don't think these errors are related to your changes because they always happen in main branch - https://ci.swift.org/view/Source%20Compatibility/job/swift-main-source-compat-suite/

In general to investigate source compatibility issues you'd have to clone `swift-source-compat-suite` repository and use `runner.py` script with `projects.json` file to trigger build of a project you are interested in. That would clone the project locally, checkout revision specified in `projects.json` and run xcodebuild or swiftpm build (depending on project configuration). `runner.py` also allows you to pass additional swift/clang flags and use custom compiler path as well.

**Comment by xedin** [comment]
Let's wait for @hborla to take a look and I think if there is nothing major this should be good to go.

File: include/swift/Sema/ConstraintSystem.h [hunk 5]
```diff
@@ -3759,14 +3759,14 @@ class ConstraintSystem {
...
+  /// types with bound generic types whose generic args are fresh type
+  /// variables.
   ///
-  /// \param type The type to open.
+  /// \param type The type on which to perform the conversion.
   ///
-  /// \returns The opened type.
-  Type openUnboundGenericTypes(Type type, ConstraintLocatorBuilder locator);
+  /// \returns The converted type.
+  Type convertInferableTypes(Type type, ConstraintLocatorBuilder locator);
```
Comment (hborla) [inline include/swift/Sema/ConstraintSystem.h:3769]: Subjective nit: I don't love the term "convert/conversion" here, because a type conversion means something different. I would rather use something like `replaceInferableTypes` or `substituteInferableTypes`. We could also go for the long form and tack on `WithTypeVar[iable]s` at the end depending on how verbose we want to be 🙂

**Review by hborla — APPROVED** [review]
LGTM! I just have one tiny opinion about a function name

File: include/swift/Sema/ConstraintSystem.h (hunk 5 above)
Comment (Jumhyn) [reply include/swift/Sema/ConstraintSystem.h:3769]: That makes sense to me re: "convert"! I'll swap this out with one of your suggestions.

**Comment by Jumhyn** [comment]
@hborla Done!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by Jumhyn** [comment]
@hborla or @xedin looks like we are good to go, mind merging? 🙂

**Comment by hborla** [comment]
@Jumhyn done, thanks again for your work on this!

**Comment by Jumhyn** [comment]
Thank you both (+ @CodaFi) for your feedback and guidance!

File: lib/Sema/TypeChecker.cpp (hunk 1 above)
Comment (viv718918) [reply lib/Sema/TypeChecker.cpp]: > Maybe not something to do here specifically, but I was really hoping we could use placeholder types to make UnboundGenericType obsolete. Wherever type resolution forms one, it could instead build a bound generic type full of placeholders. With that in mind, having them escape unopened is not so bad.
